File: packages/native-android/android/src/main/java/com/openota/runtime/BundleRollback.kt
package com.openota.runtime

import java.io.File

/**
 * Owns the single previous-generation snapshot in `rollback/`. Deliberately one generation deep,
 * not a full history: a device-side rollback is a last-resort safety net for "the last install was
 * bad," not a version-control system. Multi-generation history is a natural v2 extension (a small
 * ring of numbered directories instead of one `rollback/`) and nothing here would need to change
 * shape to support it — see the note on [RuntimeManifest] in Manifest.kt.
 */
class BundleRollback(private val storage: BundleStorage) {

    fun hasRollback(): Boolean = storage.rollbackDir.exists() && storage.manifestFile(storage.rollbackDir).exists()

    /** Moves whatever is currently active into `rollback/`, discarding any older rollback snapshot. */
    fun snapshotCurrent() {
        if (!storage.currentDir.exists() || !storage.manifestFile(storage.currentDir).exists()) {
            return
        }

        val stagedRollback = File(storage.tmpDir, "rollback-snapshot")
        storage.deleteRecursively(stagedRollback)
        storage.currentDir.copyRecursively(stagedRollback, overwrite = true)
        storage.atomicReplace(stagedRollback, storage.rollbackDir)
    }

    /** Restores `rollback/` into `current/` and clears the rollback slot (single-use). */
    fun restore(): VerifiedBundle {
        if (!hasRollback()) {
            throw NoRollbackAvailableException()
        }

        val rollbackManifest = BundleManifest.parse(storage.manifestFile(storage.rollbackDir).readText())

        val stagedCurrent = File(storage.tmpDir, "rollback-restore")
        storage.deleteRecursively(stagedCurrent)
        storage.rollbackDir.copyRecursively(stagedCurrent, overwrite = true)
        storage.atomicReplace(stagedCurrent, storage.currentDir)
        storage.deleteRecursively(storage.rollbackDir)

        return VerifiedBundle(
            packageDir = storage.currentDir,
            manifest = rollbackManifest,
            bundleFile = storage.bundleFile(storage.currentDir, rollbackManifest.bundleName),
            sha256 = rollbackManifest.sha256,
        )
    }

    fun clear() {
        storage.deleteRecursively(storage.rollbackDir)
        storage.rollbackDir.mkdirs()
    }
}

File: packages/native-ios/ios/BundleRollback.swift
import Foundation

/// Owns the single previous-generation snapshot in `rollback/` — mirrors `BundleRollback.kt`.
/// Deliberately one generation deep: a device-side rollback is a last-resort safety net, not
/// version control.
final class BundleRollback {
    private let storage: BundleStorage

    init(storage: BundleStorage) {
        self.storage = storage
    }

    func hasRollback() -> Bool {
        storage.fileExists(storage.rollbackDir, isDirectory: true) &&
            storage.fileExists(storage.manifestFile(storage.rollbackDir))
    }

    /// Moves whatever is currently active into `rollback/`, discarding any older rollback snapshot.
    func snapshotCurrent() throws {
        guard storage.fileExists(storage.currentDir, isDirectory: true),
              storage.fileExists(storage.manifestFile(storage.currentDir))
        else {
            return
        }

        let staged = storage.tmpDir.appendingPathComponent("rollback-snapshot")
        storage.deleteRecursively(staged)
        try storage.copyRecursively(from: storage.currentDir, to: staged)
        try storage.atomicReplace(source: staged, destination: storage.rollbackDir)
    }

    /// Restores `rollback/` into `current/` and clears the rollback slot (single-use).
    func restore() throws -> VerifiedBundle {
        guard hasRollback() else {
            throw OpenOTAException.noRollbackAvailable
        }

        let rollbackManifest = try BundleManifest.parse(contentsOf: storage.manifestFile(storage.rollbackDir))

        let staged = storage.tmpDir.appendingPathComponent("rollback-restore")
        storage.deleteRecursively(staged)
        try storage.copyRecursively(from: storage.rollbackDir, to: staged)
        try storage.atomicReplace(source: staged, destination: storage.currentDir)
        storage.deleteRecursively(storage.rollbackDir)

        return VerifiedBundle(
            packageDir: storage.currentDir,
            manifest: rollbackManifest,
            bundleFile: storage.bundleFile(storage.currentDir, bundleFileName: rollbackManifest.bundleName),
            sha256: rollbackManifest.sha256
        )
    }

    func clear() {
        storage.deleteRecursively(storage.rollbackDir)
        try? FileManager.default.createDirectory(at: storage.rollbackDir, withIntermediateDirectories: true)
    }
}
